### ISSUE TITLE
Gracefully handle invalid directories and files

### ISSUE THREAD

**Issue by e5150** [instaloader/instaloader#576]
**Provide us a use case of the feature**
Avoid fatal errors due to OS/filesystem character limitations.

**Describe the solution you'd like**
A wrapper could be made around `os.makedirs` and in `write_raw`, such that if an exception is raised (is it always `OSError`?), retry with a sanitised path, where everything except something like `string.ascii_letters + string.digits + '_'` (` + '/'` for dirs) is replaced with `'_'`

**Describe alternatives you've considered**
Keep adding exceptions to `_PostPathFormatter`, but having to take underlying filesystem (as per #575) into account in addition to `platform.system()` seems cumbersome.
(I'm not suggesting the existing replacement be removed, just an addition of a failsafe that covers corner cases.)

**If the feature request is accepted, would you be willing to submit a PR?**

Yes

**Comment by aandergr** [comment]
Good idea! It absolutely makes sense to retry with a sanitized path when filesystem related errors occur. Due to not removing the existing replacement, just adding a failsafe that covers the cases where the existing logic fails, backwards compatibility is assured. I also think this might help regarding our many [Windows-related bugs](https://github.com/instaloader/instaloader/issues?q=is%3Aissue+is%3Aopen+label%3AWindows), at least #185, #560, #575.

> (is it always `OSError`?)

Going through the issues, I saw mentions of `OSError` and `NotADirectoryError` that need to be handled.

**Comment by e5150** [comment]
This seems a bit more complicated than I first though, as there are a lot of places where paths are handled. If a substitution is made in a dirname, not only mkdir needs to know about it, but also the id-file handling, metadata creation. And things such as file existance check and timestamping in `download_pic()` needs to know if a file has been renamed down in `write_raw()`.
Since we can't know if a dir/file creation will be allowed before mkdir/open is called, one could do something like the following, long before the actual file creation is really needed.
```
try:
    open(path, 'w').close()
    os.unlink(path)
except OSError:
    path = sanitize(path)
```

But to avoid sprinkling a lot ugly hacks all over the code, a solution might involve a significant and well considered overhaul…
